### ISSUE TITLE
UI freezes frequently, seemingly correlated with network usage

### ISSUE THREAD

**Issue by everdred** [koreader/koreader#6421]
* KOReader version:2020.07.1
* Device: Kobo Libra H2O

#### Issue
Lately (since updating to 2020.07, I think) I've noticed the KOReader UI become frequently unresponsive during active use of the application.

#### Steps to reproduce
With continued observation over time I think it coincides with actions that use the network, such as manually pushing reading progress to cloud, going from an open book back to the file list, locking the screen via the power button, checking for an application update, etc. More basic functionality like turning pages and navigating the book work as expected. Opening a book happens quickly, however the application then becomes unresponsive to actions such as turning pages. After a period of around 15 seconds, screen taps and button presses all happen quickly, as if they had been queued.

The ssh server does not appear to be affected; it enables/disables/can be accessed instantly.

##### `crash.log` (if applicable)
[lag-log.txt](https://github.com/koreader/koreader/files/4967981/lag-log.txt)

**Comment by NiLuJe** [comment]
Can't tell much without debug logs ;).

From what little there is, possibly related to the Sync plugin?

**Comment by everdred** [comment]
Sorry, didn't see anything suggested beyond the crash.log, which I attached. Where are debug logs? Would dmesg output be helpful?

**Comment by Frenzie** [comment]
In the file manager if you go into the tools menu, in developer options you can enable debug logging. (Or you can edit the koreader.sh file to include `-d`-.)

**Comment by NiLuJe** [comment]
`dmesg` output could possibly be useful, but it's often extremely verbose and not timestamped on Kobo, which makes it harder to cross-reference ;).

**Comment by everdred** [comment]
Thanks for the responses! I found the debug option in the menu. Would verbose logging be needed?

**Comment by everdred** [comment]
Here is a few minutes of mostly opening and closing books, checking for updates a couple of times. Let me know if there's anything else you need!

[crash.log](https://github.com/koreader/koreader/files/4968367/crash.log)

One thing worth noting: when I attempt to check for updates the app freezes for a while, finally unfreezes and asks if I want to turn on wifi. I say yes, it successfully connects to my network and the update dialog disappears without telling me the result of the check.

**Comment by everdred** [comment]
I should also point out that I'm only seeing this on my Kobo; it doesn't reproduce on Android or x86_64.

**Comment by NiLuJe** [comment]
> Here is a few minutes of mostly opening and closing books, checking for updates a couple of times. Let me know if there's anything else you need!
> 
> [crash.log](https://github.com/koreader/koreader/files/4968367/crash.log)
> 

Did it stall at any particular point in time during this timeframe? if so, when? (action/timestamps).

> One thing worth noting: when I attempt to check for updates the app freezes for a while, finally unfreezes and asks if I want to turn on wifi. I say yes, it successfully connects to my network and the update dialog disappears without telling me the result of the check.

The second bit is the expected behavior. We don't/can't track the original action that requested network access, so it's not automatically resumed once connected.
That's a gnarly two-pronged issue: as I said, we don't/can't track the original action, and we can't check for a successful network connection without blocking.

The first bit leads is intriguing, though. Provided you're actually disconnected, checking that should be almost immediate. So, new questions:

* Were you disconnected at the time?
* If you weren't, how string is your WiFi network, usually, and is there anything particular about it?
* Do you have the "Restore Wi-Fi on resume" enabled?
* How do you launch KOReader?

**Comment by NiLuJe** [comment]
>     * Do you have the "Restore Wi-Fi on resume" setting enabled?

Reading the log ansdwered that for me ;p.

You do. Disable that, and I have a feeling things should get better, or at the very least more consistent.

**Comment by NiLuJe** [comment]
Other than that, the Sync stuff indeed joyously seems to be timeout'ing after 20s, which is probably what you're experiencing.

I don't use it and I have no experience with it, though.

**Comment by everdred** [comment]
> Disable that, and I have a feeling things should get better

Yep, no more lagging! But also no network, it seems. When I press the corner to manually push progress I get the 'something went wrong' error.

I should mention that the combination of settings I was using (up until just now) only started to produce this behavior with 2020.07.

**Comment by everdred** [comment]
I've been meaning to make a forum post asking what combination of network options I should select in order to save battery while keeping full functionality, even if it means waiting for occasional wifi reconnects. As of right now I have these set...

[ ] wifi connection
[ ] automatically restore after resume
action when off: turn on
[x] dismiss wifi scan popup

...but I can't seem to think my way through which supersede which, for instance whether "[ ] wifi connection" means no wifi ever, or whether having that one unchecked but "action when off: turn on" will turn that one on when needed.

**Comment by NiLuJe** [comment]
Drilling down into the Sync error, that's an `EAI_NONAME` from a `getaddrinfo` call deep inside luasocket.

In English, that's a name resolution error and timeout. Which, obviously, shouldn't happen on a sane network.

I'd ask you if you know if you're having DNS issue on your network, but I also know that the async WiFi resume can sometime botch the interface bringup, leading to.... name resolution issues ;).

**Comment by NiLuJe** [comment]
> [ ] wifi connection

That's essentially the manual "WiFi" toggle. If it's checked, WiFi is up and sucking battery.

> [ ] automatically restore after resume

That's an attempt at emulating Android behavior by attempting to restore WiFi on boot & resume from sleep. It is known to sometime outright fail, or subtly botch the bringup, for... reasons.

> action when off: turn on

That's what happens when you attempt to trigger an action requiring network while it is down. As I explained earlier, nothing will actually allow the action to go through unimpeded all on its own, though.

> [x] dismiss wifi scan popup

That just saves a click on Kobo when you have saved networks ;).

----

TL;DR: WiFi on Kobo is a wonderful Jenga tower of weird quirks, and it's already a minor miracle stuff mostly works as it does.

**Comment by everdred** [comment]
In case you'd still like answers to the previous questions...

>     Were you disconnected at the time?

As noted in the above comment, I'm never really sure if I'm connected! The settings I had been using seem to have me connected as-needed, at the very least.

>    If you weren't, how strong is your WiFi network, usually, and is there anything particular about it?

Wifi signal should be fine. I'm less than 10 feet from the router, no walls between. This is the only device regularly using my 2.4 ghz network, so I just got my phone on the network to test it. (Network and DNS work, and my phone claims to have full bars.)

>    How do you launch KOReader?

I installed it using your "one click" package, so I think that means NickelMenu?

**Comment by NiLuJe** [comment]
That said, nothing specific to 2020.07 (as compared to 2020.06) should have affected any of these WiFi quirks, so this is still mildly intriguing.

Do you happen to have paired this update with the recent Nickel FW updates?

(Not that I've noticed anything extra-weird on my Forma on that front, either).

**Comment by NiLuJe** [comment]
> I installed it using your "one click" package, so I think that means NickelMenu?

Via KFMon, yeah.

Depending on how quickly you launch KOReader after Nickel's startup, you may also be seeing interaction with Nickel's own DHCP handshake.

We attempt to take care of the most egregious issues in these cases, but there's no perfect solution.

(That's an extremely tricky corner-cases, though, and it's somewhat hard to reproduce, if ever).

**Comment by everdred** [comment]
My device is running the 6/22 firmware; not sure if that's the latest.

**Comment by NiLuJe** [comment]
There was a (couple) of July releases, so it might also be worth trying that, because Kobo's been fighting an ongoing battle with Wi-Fi on their end, too ;).

**Comment by NiLuJe** [comment]
The async restore currently *does* check if the bringup was successful, by checking for a pinging the default gateway.

We can't really make it stricter (which would be doing a DNS resolution), because there are a number of use-cases where local-only access is the desirable behavior.

So, I guess, yeah, I am still going to ask you to double-check for DSN issues on your network over the 2.4G band ;).

**Comment by NiLuJe** [comment]
If you haven't rebooted in a while, I'd do that, too, especially if you often toggle between KOReader/Nickel, as killing/restarting Nickel has a few nasty side-effects, the most visible ones being exacerbated Wi-Fi issues.

**Comment by everdred** [comment]
I can't seem to get my device to prompt me to update, and I did a sync from within Nickel... are you sure that the July firmware updates are applicable to Lirbra H2O? https://pgaskin.net/KoboStuff/kobofirmware.html doesn't list them either, but I don't know how authoritative that page is.

**Comment by everdred** [comment]
> So, I guess, yeah, I am still going to ask you to double-check for DNS issues on your network over the 2.4GHz band ;).

DNS isn't my strong suit... for testing purposes is it enough that I got my phone on the 2.4 ghz network, opened a terminal and did a few "dig doman.tld" on domains I definitely haven't visited lately (to avoid caching)?

**Comment by NiLuJe** [comment]
Nope, 4.22.15190 is definitely the last version currently offered, I just possibly remembered it as more recent than it actually is ;p.

----

Ideally, you'd do that in an SSH shell on the device itself ;p (nslookup should be available).

**Comment by everdred** [comment]
Ahh, I didn't remember nslookup. (I actually checked your "kobo-stuff" pack to see if there was a dig binary there or in busybox.) As I was saying, DNS is still somewhat mysterious to me...

```
[root@(none) ~]# nslookup example.com
Server:		192.168.0.200
Address:	192.168.0.200:53

Non-authoritative answer:
Name:	example.com
Address: 93.184.216.34

Non-authoritative answer:
Name:	example.com
Address: 2606:2800:220:1:248:1893:25c8:1946

[root@(none) ~]# nslookup asd.com
Server:		192.168.0.200
Address:	192.168.0.200:53

Non-authoritative answer:
Name:	asd.com
Address: 198.46.84.198

Non-authoritative answer:

[root@(none) ~]# nslookup goatse.cx
Server:		192.168.0.200
Address:	192.168.0.200:53

Non-authoritative answer:
Name:	goatse.cx
Address: 104.18.61.172
Name:	goatse.cx
Address: 104.18.60.172
Name:	goatse.cx
Address: 172.67.180.76

Non-authoritative answer:
Name:	goatse.cx
Address: 2606:4700:3036::6812:3dac
Name:	goatse.cx
Address: 2606:4700:3036::6812:3cac
Name:	goatse.cx
Address: 2606:4700:3034::ac43:b44c
```

**Comment by NiLuJe** [comment]
Try `sync.koreader.rocks`, but given the results, I suspect that if you do a sync right now it won't hang ;).

**Comment by everdred** [comment]
Oh... obviously, sorry.

```
[root@(none) ~]# nslookup sync.koreader.rocks
Server:		192.168.0.200
Address:	192.168.0.200:53

Non-authoritative answer:
Name:	sync.koreader.rocks
Address: 172.67.175.187
Name:	sync.koreader.rocks
Address: 104.18.45.190
Name:	sync.koreader.rocks
Address: 104.18.44.190

Non-authoritative answer:
Name:	sync.koreader.rocks
Address: 2606:4700:3037::6812:2dbe
Name:	sync.koreader.rocks
Address: 2606:4700:3030::ac43:afbb
Name:	sync.koreader.rocks
Address: 2606:4700:3032::6812:2cbe
```

**Comment by NiLuJe** [comment]
Obvious thing is obvious: I feel pretty confident almost no-one's ever checked what happens if IPv6 is actually available on your device, so I'd kill the IPv6 for your Kobos MAC on your router, on the off-chance it's screwing with us somehow ;).

**Comment by everdred** [comment]
I'm pretty confident that I've completely disabled ipv6 on my router but the nslookup output I'm seeing is the same: ipv4 and ipv6 addresses. (I know this is a case of _totally-not-your-problem_ but do you know if that's the expected behavior?)

**Comment by NiLuJe** [comment]
I"ll double-check on the Forma later, but I definitely don't see it on the H2O.

**Comment by NiLuJe** [comment]
Oops, never mind that, I was using my own busybox build w/ IPv6 disabled ;D.

I definitely see it too on the H2O, and I sure as hell don't have an IPv6 network ^^.

**Comment by everdred** [comment]
@NiLuJe I should let you know that at some point yesterday, after all the tweaking and rebooting I did, I can't get the issue to reproduce anymore! Well, I feel as if there may be the slightest touch of lag (one or two seconds) in places where it wasn't happening before, but if it is truly lag I don't think it's severe enough that I would have noticed. 

I'm happy to provide current logs (if you think that seeing the difference between now and yesterday will shed light on what was going on with all the lag yesterday), but otherwise I'm happy to go on with my life unless it crops up again. :)

**Comment by poire-z** [comment]
Might also depend on RAM usage (you can have it displayed in the footer, to correlate it to when you notice the slowdowns). cf https://github.com/koreader/koreader/issues/5998#issuecomment-602180998
So, it's all fine after a reboot or restart, but it may get worse again :/

**Comment by NiLuJe** [comment]
FWIW, I can currently fairly reliably reproduce it on my H2O while working on #6424:

If KOReader is started while network is down, if the network gets enabled by async/turn_on, name resolution is borked for the process' *lifetime* (i.e., Lua can't resolve, but my shell can, and restarting KOReader without touching the network setup *at all* magically fixes it).

If KOReader is started while network is up, much, much harder to reproduce.

At a wild guess, there's a race somewhere, or we're hitting a weird kernel/glibc/lua bug or something.

**Comment by NiLuJe** [comment]
Okay, I *seem* to have tracked it down to our use of `udhcpc`...

Switching to `dhcpcd` (like Nickel itself did, probably a while back) magically fixes all the things.

So, hopefully, it'll be fixed in #6424 ;).

**Comment by everdred** [comment]
Hey @NiLuJe, I just want to say I appreciate the time and attention you put into helping with this issue! I look forward to trying out the next release. :)

**Comment by Galunid** [comment]
@everdred Actually you can try it even now, if you're interested. Go to `Last menu item -> Update -> Settings -> Update Channel -> Development` then `Last menu item -> Update -> Check for update`. :)

**Comment by NiLuJe** [comment]
For best results, you'll have to do a full proper reboot after installing it, as some of the fixes affect the startup script ;).
